geopmdpy/geopmdpy/optimizer.py

```diff
@@ -0,0 +1,491 @@
+        result = gp_minimize(
+            func=objective,
+            dimensions=self.space,
+            n_calls=n_calls,
+            n_initial_points=n_initial_points,
+            random_state=random_state,
+            acq_func='EI'  # Expected Improvement
+        )
+
+        # Store best result
+        self.best_result = result
+
+        # Convert best coordinate to actual configuration
+        best_coordinate = [int(x) for x in result.x]
+        best_config = self.control_grid.get_config_str(best_coordinate).split('\n')
+
+        # Convert metric back if maximizing
+        best_metric = result.fun
+        if self.evaluator.maximize:
+            best_metric = -best_metric
+
+        return {
+            'best_coordinate': best_coordinate,
+            'best_metric': best_metric,
+            'best_config': best_config,
+            'n_evaluations': len(result.func_vals),
+            'optimization_result': result
+        }
+
+
+def get_parser():
+    """Create argument parser for the optimizer."""
+    parser = ArgumentParser(description="Bayesian optimization for control parameter tuning")
+
+    # Control grid options (similar to grid.py)
+    for flag, control in _CLI_FLAG_TO_CONTROL.items():
+        if flag == 'gpu_power_intel':
+            flag = 'gpu_power'
+        elif flag == 'gpu_power_nvml':
+            continue
+        flag_dash = flag.replace('_', '-')
+        parser.add_argument(
+            f"--{flag_dash}",
+            default=None,
+            dest=f'{flag}_domain',
+            help=f"Provide a grid over {control[0]} for the given domain.",
+        )
+
+    # Optimization options
+    parser.add_argument(
+        '--n-calls',
+        type=int,
+        default=50,
+        help='Number of optimization iterations (default: %(default)s)'
+    )
+
+    parser.add_argument(
+        '--n-initial-points',
+        type=int,
+        default=10,
+        help='Number of random initial evaluations (default: %(default)s)'
+    )
+
+    parser.add_argument(
+        '--metric-regex',
+        required=True,
+        help='Python-style regex to extract figure of merit from application stdout'
+    )
+
+    parser.add_argument(
+        '--minimize',
+        action='store_true',
+        help='Minimize the metric (default is to maximize)'
+    )
+
+    parser.add_argument(
+        '--random-seed',
+        type=int,
+        default=42,
+        help='Random seed for reproducibility (default: %(default)s)'
+    )
+
+    parser.add_argument(
+        '--output-best-config',
+        action='store_true',
+        help='Output the best configuration as geopmwrite commands'
+    )
+
+    # Application launch command
+    parser.add_argument(
+        'launch',
+        nargs=REMAINDER,
+        help='Command to launch application for evaluation'
+    )
+
+    return parser
+
+
+def main():
+    """Main function for the optimization tool."""
+    parser = get_parser()
+    args = parser.parse_args()
+
+    try:
```

Comment:
Add call to pio.save_control() here and restore in a finally statement.  Consider an option to just write the optimal configuration at end of run.